overlord/state/taskrunner.go

```diff
@@ -183,6 +200,30 @@ func (r *TaskRunner) AddBlocked(pred func(t *Task, running []*Task) bool) {
 	r.blocked = append(r.blocked, pred)
 }
 
+// AddHook registers a hook function for the given hook event type.
+func (r *TaskRunner) AddHook(hook HookFunc, evt HookEvent) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	r.hooks[evt] = append(r.hooks[evt], hook)
+}
+
+func (r *TaskRunner) checkExhaustion() {
+	for _, chg := range r.state.changes {
```

Comment:
even if we are in the same package we tend not to access the fields of State in taskrunner, maybe we need some helper on state to help with this code

Reply:
Good point! After passing the change of the task that has run - it was no longer neccessary to iterate all the changes and I do not access this anymore.